File: backend/src/test/java/com/codeloom/backend/exception/RestExceptionHandlerTest.java
package com.codeloom.backend.exception;

import static org.junit.jupiter.api.Assertions.*;

import org.junit.jupiter.api.*;
import org.springframework.http.HttpStatus;
import org.springframework.mock.web.MockHttpServletRequest;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.security.authentication.InsufficientAuthenticationException;

class RestExceptionHandlerTest {
    private final RestExceptionHandler restExceptionHandler = new RestExceptionHandler();
    private final MockHttpServletRequest httpServletRequest = new MockHttpServletRequest();

    @BeforeEach
    void setup() {
        httpServletRequest.setMethod("GET");
        httpServletRequest.setRequestURI("/v1/problems");
    }

    @Test
    void authenticationExceptionsReturnUnauthorized() {
        var r = restExceptionHandler.handleException(
                new InsufficientAuthenticationException("detail"), httpServletRequest);
        assertEquals(HttpStatus.UNAUTHORIZED, r.getStatusCode());
        assert r.getBody() != null;
        assertEquals("Authentication required", r.getBody().message());
        assertEquals("/v1/problems", r.getBody().path());
        assertNull(r.getBody().payload());
    }

    @Test
    void accessDeniedReturnsForbidden() {
        var r = restExceptionHandler.handleException(new AccessDeniedException("detail"), httpServletRequest);
        assertEquals(HttpStatus.FORBIDDEN, r.getStatusCode());
        assert r.getBody() != null;
        assertEquals("Access denied", r.getBody().message());
    }
}

File: backend/src/test/kotlin/com/codeloom/backend/exception/RestExceptionHandlerTest.kt
package com.codeloom.backend.exception

import org.junit.jupiter.api.Test
import org.springframework.http.HttpStatus
import org.springframework.mock.web.MockHttpServletRequest
import org.springframework.security.access.AccessDeniedException
import org.springframework.security.authentication.InsufficientAuthenticationException
import kotlin.test.assertEquals
import kotlin.test.assertNull

class RestExceptionHandlerTest {
    private val handler = RestExceptionHandler()
    private val request =
        MockHttpServletRequest().apply {
            method = "GET"
            requestURI = "/v1/problems"
        }

    @Test
    fun `authentication exceptions return standard unauthorized response`() {
        val response = handler.handleException(InsufficientAuthenticationException("internal detail"), request)

        assertEquals(HttpStatus.UNAUTHORIZED, response.statusCode)
        assertEquals(HttpStatus.UNAUTHORIZED.value(), response.body?.status)
        assertEquals("Authentication required", response.body?.message)
        assertEquals("/v1/problems", response.body?.path)
        assertNull(response.body?.payload)
    }

    @Test
    fun `access denied exceptions return standard forbidden response`() {
        val response = handler.handleException(AccessDeniedException("internal detail"), request)

        assertEquals(HttpStatus.FORBIDDEN, response.statusCode)
        assertEquals(HttpStatus.FORBIDDEN.value(), response.body?.status)
        assertEquals("Access denied", response.body?.message)
        assertEquals("/v1/problems", response.body?.path)
        assertNull(response.body?.payload)
    }
}
